website/docs/dev/cluster/migration.md

```diff
@@ -0,0 +1,310 @@
+
+The second component is separated into the following sub-components:
+
+1. ```MigrationManager```
+2. ```MigrateSessionTaskStore```
+3. ```MigrateSession```
+
+The ```MigrationManager``` is responsible for managing the active ```MigrateSession``` tasks.
+It uses ```MigrateSessionTaskStore``` to atommically add or remove new ```MigrateSession```.
+It is also responsible for ensuring that existing sessions do not conflict with session that are about to be added by checking if the referred slots in each session do not conflict.
+Finally, it provideds information on the number of ongoing migrate tasks.
+
+## Migration Data Integrity
+
+Since slot migration can be initiated while a Garnet cluster is operational, it needs to be carefully orchestrated to avoid any data integrity issues when clients attempt to write new data.
+In addition, whatever solution is put forth should not hinder data avaibility while migration is in progress.
+
+Our implementation leverages on the concept slot based sharding to ensure that keys mapping to the related slot cannot be modified while a migrate task is active.
+This is achieved by setting the slot state to ***MIGRATING*** in the source node.
+This prevents any write requests though it still allows for reads.
+Write requests can be issued towards the target node using ***ASKING***, though there are not consistency guarantees from Garnet if this option is used.
+Garnet guarantees that no keys can be lost during regular migration (i.e. with using ***ASKING***).
+
+Because Garnet operates in a multi-threaded environment, the transition from ```STABLE``` to ```MIGRATING``` needs to happen safely, so every thread has then change to observe that state change.
+This can be achieved using epoch protection.
+Specifically, when a slot transitions to a new state, the segment that implements the actual state transition will have to spin-wait after making the change and return only after all active client sessions have moved to the next epoch.
+
+An excerpt of the code related to epoch protection during slot state transition is shown below.
+Every client session, before processing an incoming packet, will first acquire the current epoch.
+This also happens for the client session that issues a slot state transition command (i.e. CLUSTER SETSLOT).
+On success of updating the local config command, the processing thread of the previous command will perform the following:
+
+1. release the current epoch (in order to avoid waiting for itselft to transition).
+2. spin-wait until all thread transition to next epoch.
+3. re-acquire the next epoch.
+
+This series of steps ensures that on return, the processing thread guarantees to the command issuer (or the method caller if state transition happens internally), that the state transition is visible to all threads that were active.
+Therefore, any active client session will process subsequent commands considering the new slot state.
+
+<details>
```

Comment:
These code blocks don't seem to render correctly in the Github markdown preview. Can you verify they will work fine in the generated documentation HTML?

Reply:
I have been using them in other parts of the website also. I checked them locally and they render fine. Will make sure again!